File: app/src/main/java/com/riprog/launcher/SettingsActivity.java
package com.riprog.launcher;

import android.app.Activity;
import android.app.UiModeManager;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.text.method.LinkMovementMethod;
import android.text.util.Linkify;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.SeekBar;
import android.widget.Switch;
import android.widget.TextView;

public class SettingsActivity extends Activity {

    private SettingsManager settingsManager;

    @Override
    protected void attachBaseContext(Context newBase) {
        SettingsManager sm = new SettingsManager(newBase);
        super.attachBaseContext(ThemeMechanism.applyThemeToContext(newBase, sm.getThemeMode()));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        settingsManager = new SettingsManager(this);
        ThemeMechanism.applyThemeMode(this, settingsManager.getThemeMode());

        Window w = getWindow();
        w.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
        w.setStatusBarColor(Color.TRANSPARENT);
        w.setNavigationBarColor(Color.TRANSPARENT);

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
            w.setDecorFitsSystemWindows(false);
        }

        FrameLayout rootContainer = new FrameLayout(this);
        rootContainer.setPadding(dpToPx(16), dpToPx(48), dpToPx(16), dpToPx(32));

        rootContainer.setOnApplyWindowInsetsListener((v, insets) -> {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.R) {
                android.graphics.Insets systemInsets = insets.getInsets(android.view.WindowInsets.Type.systemBars());
                v.setPadding(dpToPx(16), systemInsets.top + dpToPx(16), dpToPx(16), systemInsets.bottom + dpToPx(16));
            } else {
                v.setPadding(dpToPx(16), insets.getSystemWindowInsetTop() + dpToPx(16), dpToPx(16), insets.getSystemWindowInsetBottom() + dpToPx(16));
            }
            return insets;
        });

        ScrollView scrollView = new ScrollView(this);
        scrollView.setBackground(ThemeUtils.getGlassDrawable(this, settingsManager));
        scrollView.setVerticalScrollBarEnabled(false);
        rootContainer.addView(scrollView);

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(dpToPx(24), dpToPx(32), dpToPx(24), dpToPx(32));
        scrollView.addView(root);

        int adaptiveColor = ThemeUtils.getAdaptiveColor(this, settingsManager, true);

        LinearLayout titleLayout = new LinearLayout(this);
        titleLayout.setOrientation(LinearLayout.HORIZONTAL);
        titleLayout.setGravity(Gravity.CENTER_VERTICAL);
        titleLayout.setPadding(0, 0, 0, dpToPx(32));

        ImageView titleIcon = new ImageView(this);
        titleIcon.setImageResource(R.drawable.ic_settings);
        titleIcon.setColorFilter(adaptiveColor);
        LinearLayout.LayoutParams iconParams = new LinearLayout.LayoutParams(dpToPx(32), dpToPx(32));
        iconParams.rightMargin = dpToPx(12);
        titleLayout.addView(titleIcon, iconParams);

        TextView title = new TextView(this);
        title.setText(R.string.title_settings);
        title.setTextSize(32);
        title.setTypeface(Typeface.create("sans-serif-light", Typeface.NORMAL));
        title.setTextColor(adaptiveColor);
        titleLayout.addView(title);
        root.addView(titleLayout);

        addCategoryHeader(root, getString(R.string.category_home), R.drawable.ic_layout);
        addFreeformSetting(root);
        addHideLabelsSetting(root);

        addCategoryHeader(root, getString(R.string.category_appearance), R.drawable.ic_wallpaper);
        addThemeSetting(root);
        addLiquidGlassSetting(root);
        addDarkenWallpaperSetting(root);
        addScaleSetting(root);

        addCategoryHeader(root, getString(R.string.category_about), R.drawable.ic_info);

        TextView aboutContent = new TextView(this);
        aboutContent.setText(R.string.about_content);
        aboutContent.setTextColor(adaptiveColor & 0xBBFFFFFF);
        aboutContent.setTextSize(14);
        aboutContent.setPadding(0, 0, 0, dpToPx(32));
        Linkify.addLinks(aboutContent, Linkify.WEB_URLS);
        aboutContent.setMovementMethod(LinkMovementMethod.getInstance());
        root.addView(aboutContent);

        setContentView(rootContainer);
    }

    private void addCategoryHeader(LinearLayout parent, String title, int iconRes) {
        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.HORIZONTAL);
        layout.setGravity(Gravity.CENTER_VERTICAL);
        layout.setPadding(0, dpToPx(24), 0, dpToPx(12));

        int adaptiveColor = ThemeUtils.getAdaptiveColor(this, settingsManager, true);

        if (iconRes != 0) {
            ImageView icon = new ImageView(this);
            icon.setImageResource(iconRes);
            icon.setColorFilter(adaptiveColor);
            LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(dpToPx(20), dpToPx(20));
            lp.rightMargin = dpToPx(8);
            layout.addView(icon, lp);
        }

        TextView tv = new TextView(this);
        tv.setText(title);
        tv.setTextSize(14);
        tv.setTypeface(null, Typeface.BOLD);
        tv.setTextColor(adaptiveColor);
        tv.setAllCaps(true);
        layout.addView(tv);

        parent.addView(layout);
    }

    private void addThemeSetting(LinearLayout parent) {
        LinearLayout item = new LinearLayout(this);
        item.setOrientation(LinearLayout.VERTICAL);
        item.setPadding(dpToPx(16), dpToPx(16), dpToPx(16), dpToPx(16));
        ThemeMechanism.applySettingItemStyle(this, item, settingsManager);

        int adaptiveColor = ThemeUtils.getAdaptiveColor(this, settingsManager, true);

        TextView titleView = new TextView(this);
        titleView.setText(R.string.setting_theme_mode);
        titleView.setTextSize(18);
        titleView.setTextColor(adaptiveColor);
        item.addView(titleView);

        String[] modes = {getString(R.string.theme_system), getString(R.string.theme_light), getString(R.string.theme_dark)};
        String[] values = {"system", "light", "dark"};
        String current = settingsManager.getThemeMode();

        LinearLayout optionsLayout = new LinearLayout(this);
        optionsLayout.setOrientation(LinearLayout.HORIZONTAL);
        optionsLayout.setPadding(0, dpToPx(8), 0, 0);

        for (int i = 0; i < modes.length; i++) {
            final int index = i;
            TextView option = new TextView(this);
            option.setText(modes[i]);
            option.setPadding(dpToPx(12), dpToPx(8), dpToPx(12), dpToPx(8));
            option.setTextSize(14);

            boolean isSelected = values[i].equals(current);
            option.setTextColor(isSelected ? adaptiveColor : adaptiveColor & 0xBBFFFFFF);

            if (isSelected) {
                GradientDrawable gd = new GradientDrawable();
                gd.setColor(getColor(R.color.search_background));
                gd.setCornerRadius(dpToPx(8));
                option.setBackground(gd);
            }

            option.setOnClickListener(v -> {
                settingsManager.setThemeMode(values[index]);
                ThemeMechanism.applyThemeMode(this, values[index]);
                recreate();
            });

            LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1);
            optionsLayout.addView(option, lp);
            option.setGravity(Gravity.CENTER);
        }
        item.addView(optionsLayout);

        LinearLayout.LayoutParams itemLp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        itemLp.bottomMargin = dpToPx(8);
        parent.addView(item, itemLp);
    }

    private void addFreeformSetting(LinearLayout parent) {
        addToggleSetting(parent, R.string.setting_freeform, R.string.setting_freeform_summary,
                settingsManager.isFreeformHome(), (isChecked) -> settingsManager.setFreeformHome(isChecked));
    }

    private void addHideLabelsSetting(LinearLayout parent) {
        addToggleSetting(parent, R.string.setting_hide_labels, R.string.setting_hide_labels_summary,
                settingsManager.isHideLabels(), (isChecked) -> settingsManager.setHideLabels(isChecked));
    }

    private void addLiquidGlassSetting(LinearLayout parent) {
        addToggleSetting(parent, R.string.setting_liquid_glass, R.string.setting_liquid_glass_summary,
                settingsManager.isLiquidGlass(), (isChecked) -> {
                    settingsManager.setLiquidGlass(isChecked);
                    recreate();
                });
    }

    private void addDarkenWallpaperSetting(LinearLayout parent) {
        addToggleSetting(parent, R.string.setting_darken_wallpaper, R.string.setting_darken_wallpaper_summary,
                settingsManager.isDarkenWallpaper(), (isChecked) -> settingsManager.setDarkenWallpaper(isChecked));
    }

    private void addToggleSetting(LinearLayout parent, int titleRes, int summaryRes, boolean isChecked, OnCheckedChangeListener listener) {
        LinearLayout item = new LinearLayout(this);
        item.setOrientation(LinearLayout.HORIZONTAL);
        item.setGravity(Gravity.CENTER_VERTICAL);
        item.setPadding(dpToPx(16), dpToPx(16), dpToPx(16), dpToPx(16));
        ThemeMechanism.applySettingItemStyle(this, item, settingsManager);

        LinearLayout textLayout = new LinearLayout(this);
        textLayout.setOrientation(LinearLayout.VERTICAL);
        LinearLayout.LayoutParams textParams = new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1);
        item.addView(textLayout, textParams);

        int adaptiveColor = ThemeUtils.getAdaptiveColor(this, settingsManager, true);

        TextView titleView = new TextView(this);
        titleView.setText(titleRes);
        titleView.setTextSize(18);
        titleView.setTextColor(adaptiveColor);
        textLayout.addView(titleView);

        TextView summaryView = new TextView(this);
        summaryView.setText(summaryRes);
        summaryView.setTextSize(14);
        summaryView.setTextColor(adaptiveColor & 0xBBFFFFFF);
        textLayout.addView(summaryView);

        Switch toggle = new Switch(this);
        toggle.setChecked(isChecked);
        toggle.setClickable(false);
        item.addView(toggle);

        item.setOnClickListener(v -> {
            boolean newState = !toggle.isChecked();
            toggle.setChecked(newState);
            listener.onCheckedChanged(newState);
        });

        LinearLayout.LayoutParams itemLp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        itemLp.bottomMargin = dpToPx(8);
        parent.addView(item, itemLp);
    }

    private void addScaleSetting(LinearLayout parent) {
        LinearLayout item = new LinearLayout(this);
        item.setOrientation(LinearLayout.VERTICAL);
        item.setPadding(dpToPx(16), dpToPx(16), dpToPx(16), dpToPx(16));
        ThemeMechanism.applySettingItemStyle(this, item, settingsManager);

        int adaptiveColor = ThemeUtils.getAdaptiveColor(this, settingsManager, true);

        TextView titleView = new TextView(this);
        titleView.setText(R.string.setting_scale);
        titleView.setTextSize(18);
        titleView.setTextColor(adaptiveColor);
        item.addView(titleView);

        SeekBar seekBar = new SeekBar(this);
        seekBar.setMax(100);
        float currentScale = settingsManager.getIconScale();
        seekBar.setProgress((int) ((currentScale - 0.5f) / 1.0f * 100));

        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                float scale = 0.5f + (progress / 100.0f) * 1.0f;
                settingsManager.setIconScale(scale);
            }
            @Override public void onStartTrackingTouch(SeekBar seekBar) {}
            @Override public void onStopTrackingTouch(SeekBar seekBar) {}
        });

        item.addView(seekBar);

        TextView description = new TextView(this);
        description.setText(R.string.setting_scale_summary);
        description.setTextSize(12);
        description.setTextColor(adaptiveColor & 0xBBFFFFFF);
        item.addView(description);

        LinearLayout.LayoutParams itemLp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        itemLp.bottomMargin = dpToPx(8);
        parent.addView(item, itemLp);
    }

    private int dpToPx(int dp) {
        return (int) android.util.TypedValue.applyDimension(
                android.util.TypedValue.COMPLEX_UNIT_DIP, dp, getResources().getDisplayMetrics());
    }

    private interface OnCheckedChangeListener {
        void onCheckedChanged(boolean isChecked);
    }
}

File: app/src/main/java/com/riprog/launcher/SettingsActivity.kt
package com.riprog.launcher
import android.app.Activity
import android.app.UiModeManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.graphics.toArgb
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
class SettingsActivity : ComponentActivity() {
    private lateinit var settingsManager: SettingsManager
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        settingsManager = SettingsManager(this)
        enableEdgeToEdge()
        applyThemeMode(settingsManager.themeMode)
        if (settingsManager.isLiquidGlass()) {
            ThemeUtils.applyWindowBlur(window, true)
        }
        setContent {
            SettingsScreen(
                settingsManager = settingsManager,
                onClose = { finish() },
                onRecreate = { recreate() }
            )
        }
    }
    private fun applyThemeMode(mode: String?) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val uiModeManager = getSystemService(Context.UI_MODE_SERVICE) as UiModeManager
            var nightMode = UiModeManager.MODE_NIGHT_AUTO
            if ("light" == mode) nightMode = UiModeManager.MODE_NIGHT_NO
            else if ("dark" == mode) nightMode = UiModeManager.MODE_NIGHT_YES
            if (uiModeManager.nightMode != nightMode) {
                uiModeManager.setApplicationNightMode(nightMode)
            }
        }
    }
}
@Composable
fun SettingsScreen(
    settingsManager: SettingsManager,
    onClose: () -> Unit,
    onRecreate: () -> Unit
) {
    val context = LocalContext.current
    val scrollState = rememberScrollState()
    val isLiquidGlass = settingsManager.isLiquidGlass()
    val adaptiveColorInt = ThemeUtils.getAdaptiveColor(context, settingsManager, true)
    val adaptiveColor = Color(adaptiveColorInt)
    val secondaryColor = adaptiveColor.copy(alpha = 0.7f)
    val backgroundColorInt = if (isLiquidGlass) {
        context.getColor(R.color.background)
    } else {
        if (context.resources.configuration.uiMode and android.content.res.Configuration.UI_MODE_NIGHT_MASK ==
            android.content.res.Configuration.UI_MODE_NIGHT_YES) android.graphics.Color.BLACK else android.graphics.Color.WHITE
    }
    val backgroundColor = Color(backgroundColorInt)
    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(backgroundColor)
            .statusBarsPadding()
            .navigationBarsPadding()
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(scrollState)
                .padding(horizontal = 24.dp)
        ) {
            Spacer(modifier = Modifier.height(64.dp))
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.padding(bottom = 24.dp)
            ) {
                Icon(
                    painter = painterResource(id = R.drawable.ic_settings),
                    contentDescription = null,
                    tint = adaptiveColor,
                    modifier = Modifier.size(32.dp)
                )
                Spacer(modifier = Modifier.width(12.dp))
                Text(
                    text = stringResource(id = R.string.title_settings),
                    fontSize = 32.sp,
                    fontWeight = FontWeight.Bold,
                    color = adaptiveColor
                )
            }
            CategoryHeader(
                title = stringResource(id = R.string.category_home),
                icon = painterResource(id = R.drawable.ic_layout),
                color = adaptiveColor
            )
            var freeform by remember { mutableStateOf(settingsManager.isFreeformHome()) }
            ToggleSetting(
                title = stringResource(id = R.string.setting_freeform),
                summary = stringResource(id = R.string.setting_freeform_summary),
                isChecked = freeform,
                onCheckedChange = {
                    freeform = it
                    settingsManager.setFreeformHome(it)
                },
                textColor = adaptiveColor,
                secondaryColor = secondaryColor
            )
            var hideLabels by remember { mutableStateOf(settingsManager.isHideLabels()) }
            ToggleSetting(
                title = stringResource(id = R.string.setting_hide_labels),
                summary = stringResource(id = R.string.setting_hide_labels_summary),
                isChecked = hideLabels,
                onCheckedChange = {
                    hideLabels = it
                    settingsManager.setHideLabels(it)
                },
                textColor = adaptiveColor,
                secondaryColor = secondaryColor
            )
            CategoryHeader(
                title = stringResource(id = R.string.category_appearance),
                icon = painterResource(id = R.drawable.ic_wallpaper),
                color = adaptiveColor
            )
            ThemeSetting(
                currentMode = settingsManager.themeMode ?: "system",
                onModeSelected = {
                    settingsManager.themeMode = it
                    onRecreate()
                },
                textColor = adaptiveColor,
                secondaryColor = secondaryColor
            )
            ToggleSetting(
                title = stringResource(id = R.string.setting_liquid_glass),
                summary = stringResource(id = R.string.setting_liquid_glass_summary),
                isChecked = isLiquidGlass,
                onCheckedChange = {
                    settingsManager.setLiquidGlass(it)
                    onRecreate()
                },
                textColor = adaptiveColor,
                secondaryColor = secondaryColor
            )
            var darkenWallpaper by remember { mutableStateOf(settingsManager.isDarkenWallpaper()) }
            ToggleSetting(
                title = stringResource(id = R.string.setting_darken_wallpaper),
                summary = stringResource(id = R.string.setting_darken_wallpaper_summary),
                isChecked = darkenWallpaper,
                onCheckedChange = {
                    darkenWallpaper = it
                    settingsManager.setDarkenWallpaper(it)
                },
                textColor = adaptiveColor,
                secondaryColor = secondaryColor
            )
            var iconScale by remember { mutableFloatStateOf(settingsManager.iconScale) }
            ScaleSetting(
                scale = iconScale,
                onScaleChange = {
                    iconScale = it
                    settingsManager.iconScale = it
                },
                textColor = adaptiveColor,
                secondaryColor = secondaryColor
            )
            CategoryHeader(
                title = stringResource(id = R.string.category_about),
                icon = painterResource(id = R.drawable.ic_info),
                color = adaptiveColor
            )
            Text(
                text = stringResource(id = R.string.about_content),
                color = secondaryColor,
                fontSize = 14.sp,
                modifier = Modifier.padding(start = 16.dp, end = 16.dp, bottom = 32.dp)
            )
            Spacer(modifier = Modifier.height(32.dp))
        }
        IconButton(
            onClick = onClose,
            modifier = Modifier
                .align(Alignment.TopEnd)
                .padding(16.dp)
                .size(48.dp)
        ) {
            Icon(
                imageVector = Icons.Default.Close,
                contentDescription = null,
                tint = adaptiveColor.copy(alpha = 0.6f),
                modifier = Modifier.size(24.dp)
            )
        }
    }
}
@Composable
fun CategoryHeader(title: String, icon: Painter, color: Color) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier.padding(start = 8.dp, top = 32.dp, bottom = 16.dp)
    ) {
        Icon(
            painter = icon,
            contentDescription = null,
            tint = color,
            modifier = Modifier.size(24.dp)
        )
        Spacer(modifier = Modifier.width(12.dp))
        Text(
            text = title,
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            color = color
        )
    }
}
@Composable
fun ToggleSetting(
    title: String,
    summary: String,
    isChecked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
    textColor: Color,
    secondaryColor: Color
) {
    Surface(
        onClick = { onCheckedChange(!isChecked) },
        color = Color.Transparent,
        shape = RoundedCornerShape(12.dp),
        modifier = Modifier
            .fillMaxWidth()
            .padding(bottom = 8.dp)
    ) {
        Row(
            modifier = Modifier.padding(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Text(text = title, fontSize = 18.sp, color = textColor)
                Text(text = summary, fontSize = 14.sp, color = secondaryColor)
            }
            Switch(checked = isChecked, onCheckedChange = onCheckedChange)
        }
    }
}
@Composable
fun ThemeSetting(
    currentMode: String,
    onModeSelected: (String) -> Unit,
    textColor: Color,
    secondaryColor: Color
) {
    val context = LocalContext.current
    val modes = arrayOf(
        stringResource(id = R.string.theme_system),
        stringResource(id = R.string.theme_light),
        stringResource(id = R.string.theme_dark)
    )
    val values = arrayOf("system", "light", "dark")
    Surface(
        color = Color.Transparent,
        shape = RoundedCornerShape(12.dp),
        modifier = Modifier
            .fillMaxWidth()
            .padding(bottom = 8.dp)
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(text = stringResource(id = R.string.setting_theme_mode), fontSize = 18.sp, color = textColor)
            Spacer(modifier = Modifier.height(8.dp))
            Row(modifier = Modifier.fillMaxWidth()) {
                values.forEachIndexed { index, value ->
                    val isSelected = value == currentMode
                    Box(
                        contentAlignment = Alignment.Center,
                        modifier = Modifier
                            .weight(1f)
                            .padding(horizontal = 4.dp)
                            .background(
                                color = if (isSelected) Color(context.getColor(R.color.search_background)) else Color.Transparent,
                                shape = RoundedCornerShape(8.dp)
                            )
                            .clickable { onModeSelected(value) }
                            .padding(vertical = 8.dp)
                    ) {
                        Text(
                            text = modes[index],
                            fontSize = 14.sp,
                            color = if (isSelected) textColor else secondaryColor
                        )
                    }
                }
            }
        }
    }
}
@Composable
fun ScaleSetting(
    scale: Float,
    onScaleChange: (Float) -> Unit,
    textColor: Color,
    secondaryColor: Color
) {
    Surface(
        color = Color.Transparent,
        shape = RoundedCornerShape(12.dp),
        modifier = Modifier
            .fillMaxWidth()
            .padding(bottom = 8.dp)
    ) {
        Column(modifier = Modifier.padding(16.dp)) {
            Text(text = stringResource(id = R.string.setting_scale), fontSize = 18.sp, color = textColor)
            Slider(
                value = scale,
                onValueChange = onScaleChange,
                valueRange = 0.5f..1.5f
            )
            Text(text = stringResource(id = R.string.setting_scale_summary), fontSize = 12.sp, color = secondaryColor)
        }
    }
}
